### PR TITLE
WIP: Fixed export param being ignored during API list operations.

### LINKED ISSUE #49497: API Export Parameter Ignored for List
/kind bug

**What happened**:

When running kubectl get secrets --export -o yaml, no metadata is stripped and no export logic is applied. The same is true for all API types.

**What you expected to happen**:

Behavior should match getting individual secrets, (kubectl get secret foo --export -o yaml) and export logic should be applied even when listing resources.

**How to reproduce it (as minimally and precisely as possible)**:

Using a hack/local-up-cluster.sh cluster:

```
echo -n "admin" > ./username.txt
echo -n "1f2d1e2e67df" > ./password.txt
kubectl create secret generic db-user-pass --from-file=./username.txt --from-file=./password.txt
kubectl get secrets -o yaml
```

You will see various fields in the output such as creationTimestamp, namespace, UID, all of which should be filtered out when exporting an object. (which they are when doing an individual GET)

**Anything else we need to know?**:

Via @deads2k the export option may have been missed on ListOptions, or we should be passing ExportOptions everywhere we're passing ListOptions.

Working on a PR to fix this issue.

**Environment**:
- Kubernetes version (use `kubectl version`):
  ```
  Client Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.0-alpha.2.709+6fdf0e4157c76e-dirty", GitCommit:"6fdf0e4157c76eeaf7619e8411ee7231305f7d87", GitTreeState:"dirty", BuildDate:"2017-07-24T13:51:15Z", GoVersion:"go1.8.1", Compiler:"gc", Platform:"linux/amd64"}
  Server Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.0-alpha.2.709+6fdf0e4157c76e-dirty", GitCommit:"6fdf0e4157c76eeaf7619e8411ee7231305f7d87", GitTreeState:"dirty", BuildDate:"2017-07-24T13:51:15Z", GoVersion:"go1.8.1", Compiler:"gc", Platform:"linux/amd64"}
  ```
- Cloud provider or hardware configuration**: N/A
- OS (e.g. from /etc/os-release):  Fedora 26
- Kernel (e.g. `uname -a`): Linux wrx 4.11.9-300.fc26.x86_64 #1 SMP Wed Jul 5 16:21:56 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

### PR DESCRIPTION
**Pull request by dgoodwin**
**What this PR does / why we need it**:

Today the export query param, most likely exposed via kubectl get --export, is applied when you fetch a specific object, but ignored if list all. Arguably this is the most important use case when exporting objects.

**Which issue this PR fixes** *: fixes #49497

**Special notes for your reviewer**:

Previously the Exporter.Export method was linked with Getter.Get and ran it internally on a given object name. In order to re-use this logic for listing as well, the best approach appeared to be separating these two and having Exporter.Export just operate on a pre-fetched runtime.Object, rather than looking it up by name. handlers/rest.go becomes responsible for parsing out ExportOptions from the query, doing it's Get/List normally, and then deciding if it can perform the export or not.

Based on feedback in the linked PR this also hooked up export=1 in the Accept header parsed during content negotiation, but looked to be unused. Both the export=true query param and the export=1 in accept header are now respected, processed by the same block of code (transformResponseObject), for get, list, and watch.

transformResponseObject is now also used in watch servers clearing out a TODO and allowing the above to be possible. To do this it had to be modified to separate the transformation of the response object from writing it out. (as this is different for watch servers vs regular API calls)

**Release note**:

```release-note
Fixed export parameter being ignored when listing API resources.
```

### CODE REVIEW

**Comment by dgoodwin** [comment]
This is rough WIP, I'm still working on watch, there might be a issue with namespace not being stripped during export in some cases, there's debug output floating around, and naturally all commits will be squashed before final form. However just wanted to get this out as it's been quite a bit of work and if anyone was willing to give early feedback it would be immensely helpful!

**Comment by dgoodwin** [comment]
CC @kubernetes/sig-api-machinery-misc

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go [hunk 1]
```diff
@@ -295,6 +300,18 @@ func ListResource(r rest.Lister, rw rest.Watcher, scope RequestScope, forceWatch
...
 		ctx = request.WithNamespace(ctx, namespace)
 
+		exportOptions := metav1.ExportOptions{}
+		if err := metainternalversion.ParameterCodec.DecodeParameters(req.URL.Query(), scope.MetaGroupVersion, &exportOptions); err != nil {
+			err = errors.NewBadRequest(err.Error())
+			scope.err(err, w, req)
+			return
+		}
+		if exportOptions.Export && e == nil {
+			err = errors.NewBadRequest(fmt.Sprintf("export of %q is not supported", scope.Resource.Resource))
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: you don't need the exporter here, right? if export=true routed to List, export is not supported

Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: oh, I see... you're adding export behavior to List

Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: > oh, I see... you're adding export behavior to List

Which is a natural use-case for export

Comment (dgoodwin) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go]: 👍

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go [hunk 2]
```diff
@@ -269,6 +287,15 @@ func (s *WatchServer) HandleWS(ws *websocket.Conn) {
 			}
 			obj := event.Object
 			s.Fixup(obj)
+			if s.ExportOptions.Export && s.Exporter != nil {
+				var err error
+				// TODO: Passing nil for context here as there is no request from which to create one:
+				// ctx *may* be completely unused in all exporters?
+				obj, err = s.Exporter.Export(nil, event.Object, s.ExportOptions)
```
Comment (dgoodwin) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go]: Anyone have thoughts on this situation? I'd like to apply export logic as we do up in ServeHTTP or over in the rest.go endpoint handlers, but we have no such request scope here.

Most exporters don't seem to use context, but for example the certificates strategy passes through to this which does: https://github.com/dgoodwin/kubernetes/blob/2568a92119833dcb76628fcf8d3a46e37abc06fe/pkg/registry/certificates/certificates/strategy.go#L66-L66

**Comment by dgoodwin** [comment]
Updated such that watch servers now go through transformResponseObject for both ServeHTTP and ServeWS, clearing out some TODOs in the process. This part in particular can be seen in c4eef0f101314d1ba8a9d934289334461d3a5d32, it required separating the writing from the transforming.

CC @smarterclayton @sttts if anyone has feedback, probably close to dropping WIP in subject at this point.

**Comment by dgoodwin** [comment]
Rebased onto master. Can we get this rolling again? As far as I know this change does the trick as is, just needs commits squashed once I can get some confidence it's in usable shape.

**Comment by dgoodwin** [comment]
Going to close this for now, there appears to be growing sentiment that export is not the best approach and we should just use full API objects and filter down on import, as this is more flexible and safer.

**Pull request closed without merging**